File: src/main/java/moe/gensoukyo/npcspawner/Region3d.java
package moe.gensoukyo.npcspawner;

import java.util.Objects;

public class Region3d {

    SimpleVec3d p1, p2, dif;

    public Region3d(double x1, double y1, double z1, double x2, double y2, double z2) {
        this.p1 = new SimpleVec3d(x1, y1, z1);
        this.p2 = new SimpleVec3d(x2, y2, z2);
        fixVec();
    }

    public Region3d(SimpleVec3d pos1, SimpleVec3d pos2) {
        this.p1 = pos1;
        this.p2 = pos2;
        fixVec();
    }

    public boolean isVecInRegion(SimpleVec3d vec) {
        if (dif == null) dif = p2.subtract(p1);
        SimpleVec3d a = vec.subtract(p1);
        SimpleVec3d n = dif.subtract(a);
        return a.x > 0 && a.y > 0 && a.z > 0 && n.x > 0 && n.y > 0 && n.z > 0;
    }

    public boolean isRegionCoincide(Region3d o) {
        return p1.x < o.p2.x && p2.x > o.p1.x
                && p1.y < o.p2.y && p2.y > o.p1.y
                && p1.z < o.p2.z && p2.z > o.p1.z;
    }

    private void fixVec() {
        double[] min = new double[3];
        double[] max = new double[3];
        findMinMax(p1.x, p2.x, min, max, 0);
        findMinMax(p1.y, p2.y, min, max, 1);
        findMinMax(p1.z, p2.z, min, max, 2);
        p1 = new SimpleVec3d(min[0], min[1], min[2]);
        p2 = new SimpleVec3d(max[0], max[1], max[2]);
    }

    private void findMinMax(double a, double b, double[] min, double[] max, int i) {
        if (a <= b) {min[i] = a; max[i] = b;} else {min[i] = b; max[i] = a;}
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Region3d region3d = (Region3d) o;
        return Objects.equals(p1, region3d.p1) && Objects.equals(p2, region3d.p2);
    }

    @Override
    public int hashCode() {
        return Objects.hash(p1, p2);
    }

    @Override
    public String toString() {
        return "Region3d{" +
                "p1=" + p1 +
                ", p2=" + p2 +
                '}';
    }
}

File: src/main/kotlin/moe/gensoukyo/npcspawner/Region3d.kt
package moe.gensoukyo.npcspawner

import net.minecraft.util.math.Vec3d

class Region3d {
    @JvmField var vec1: Vec3d
    @JvmField val vec2: Vec3d

    constructor(v1: Vec3d, v2: Vec3d) {
        val fx = v1.x < v2.x
        val fy = v1.y < v2.y
        val fz = v1.z < v2.z

        if (fx && fy && fz) {
            vec1 = v1
            vec2 = v2
        } else {
            vec1 = Vec3d(if (fx) v1.x else v2.x, if (fy) v1.y else v2.y, if (fz) v1.z else v2.z)
            vec2 = Vec3d(if (fx) v2.x else v1.x, if (fy) v2.y else v1.y, if (fz) v2.z else v1.z)
        }
    }

    constructor(x1: Double, y1: Double, z1: Double, x2: Double, y2: Double, z2: Double){
        val fx = x1 < x2
        val fy = y1 < y2
        val fz = z1 < z2
        vec1 = Vec3d(if (fx) x1 else x2, if (fy) y1 else y2, if (fz) z1 else z2)
        vec2 = Vec3d(if (fx) x2 else x1, if (fy) y2 else y1, if (fz) z2 else z1)
    }

    constructor(x1: Double, z1: Double, x2: Double, z2: Double) : this(x1, 0.0, z1, x2, 255.0, z2)
    constructor(vec1: Vec2d, vec2: Vec2d) : this(vec1.x, vec1.y, vec2.x, vec2.y)

    fun isVecInRegion(vec2d: Vec2d) = vecIn2d(vec2d.x, vec2d.y)

    fun isVecInRegion(v: Vec3d) = vecIn2d(v.x, v.z) && v.y > vec1.y && v.y < vec2.y

    private fun vecIn2d(x: Double, z: Double) = x > vec1.x && x < vec2.x && z > vec1.z && z < vec2.z

    fun isCoincideWith(region: Region2d): Boolean {
        val a1 = vec1
        val a2 = vec2
        val b1 = region.vec1
        val b2 = region.vec2
        return a2.x > b1.x && a1.x < b2.x && a1.z < b2.y && a2.z > b1.y
    }

    fun isCoincideWith(region: Region3d): Boolean {
        val a1 = vec1
        val a2 = vec2
        val b1: Vec3d = region.vec1
        val b2: Vec3d = region.vec2
        return a2.x > b1.x && a1.x < b2.x && a2.y > b1.y && a1.y < b2.y && a2.z > b1.z && a1.z < b2.z
    }
}